controllers/kubeadmconfig_controller.go

```diff
@@ -263,6 +267,83 @@ func (r *KubeadmConfigReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, re
 	return ctrl.Result{}, nil
 }
 
+// reconcileDiscovery ensure that config.JoinConfiguration.Discovery is properly set for the joining node.
+// The implementation func respect user provided discovery configurations, but in case some of them are missing, a valid BootstrapToken object
+// is automatically injected into config.JoinConfiguration.Discovery.
+// This allows to simplify configuration UX, by providing the option to delegate to CABPK the configuration of kubeadm join discovery.
+func (r *KubeadmConfigReconciler) reconcileDiscovery(cluster *capiv1alpha2.Cluster, config *cabpkv1alpha2.KubeadmConfig) error {
+	log := r.Log.WithValues("kubeadmconfig", fmt.Sprintf("%s/%s", config.Namespace, config.Name))
+
+	// if config already contains a file discovery configuration, respect it without further validations
+	if config.Spec.JoinConfiguration.Discovery.File != nil {
+		return nil
+	}
+
+	// otherwise it is necessary to ensure token discovery is properly configured
+	if config.Spec.JoinConfiguration.Discovery.BootstrapToken == nil {
+		config.Spec.JoinConfiguration.Discovery.BootstrapToken = &kubeadmv1beta1.BootstrapTokenDiscovery{}
+	}
+
+	// if BootstrapToken already contains an APIServerEndpoint, respect it; otherwise inject the APIServerEndpoint endpoint defined in cluster status
+	//TODO(fp) might be we want to validate user provided APIServerEndpoint and warn/error if it doesn't match the api endpoint defined at cluster level
+	apiServerEndpoint := config.Spec.JoinConfiguration.Discovery.BootstrapToken.APIServerEndpoint
+	if apiServerEndpoint == "" {
+		if len(cluster.Status.APIEndpoints) == 0 {
+			return errors.New("Control plane already exists for the cluster but cluster.Status.APIEndpoints are not set")
```

Comment:
I would expect to return requeue after some duration instead of returning an error here. My thinking is that if the status.APIEndpoints doesn't exist then it means we're waiting on cluster-api to write that value.

Reply:
Discussed offline. This is returning an error because it felt awkward to return (ctrl.Result{}, error). I suggested either returning (ctrl.Result{}, error) or custom errors and handling them in Reconcile.

Reply:
There are previous art in cluster API of  nested reconcile funcs returning an error, and there is also an error type that specifically triggers requeue